<a href="https://colab.research.google.com/github/BhavyaSrinivas2003/CreditCardFraudDetect/blob/main/SVMCreditCardFraudDetect.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**GATHERING DATA from kaggle**

In [ ]:
#Installing kaggle module
!pip install kaggle

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**Command Notations:**
* mkdir--make Directory
* cp-- copy
* chmod--change mode
*json--JavaScriptObjectNotation

In [ ]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

Dataset Command Path

In [ ]:
!kaggle datasets download -d mlg-ulb/creditcardfraud

 86% 57.0M/66.0M [00:00<00:00, 93.6MB/s]
100% 66.0M/66.0M [00:00<00:00, 81.1MB/s]


In [ ]:
#extracting zip file
from zipfile import ZipFile
dataset="/content/creditcardfraud.zip"
with ZipFile(dataset,'r')as zip:
  zip.extractall()
  print("The data is extracted")

The data is extracted


In [ ]:
print(dataset)

/content/creditcardfraud.zip


In [ ]:
import pandas as pd

In [ ]:
Data=pd.read_csv("/content/creditcard.csv")
Data

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

**PREPARING THE DATA**

In [ ]:
#importing standard libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
#importing svm model
from sklearn import svm
from sklearn.metrics import accuracy_score

In [ ]:
Dataset=pd.read_csv("/content/creditcard.csv")
print(Dataset)

            Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.01830

In [ ]:
Dataset.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
Dataset.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [ ]:
Dataset.shape

(284807, 31)

**FINDING NULL VALUES**

In [ ]:
Dataset.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

**Checking wheather the data is balanced or unbalanced**

In [ ]:
Dataset["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Hence the data is clearly unbalanced. and we need to convert it into Balanced format.

**CLASSIFYING VALID USER AND FRAUD USER**

In [ ]:
valid=Dataset[Dataset.Class==0]
valid

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284315 rows x 31 columns]

In [ ]:
fraud=Dataset[Dataset.Class==1]
fraud

Time        V1        V2        V3        V4        V5        V6  \
541        406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623        472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920      4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108      6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329      7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
541    -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623     0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920    0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108   -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329    1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
541     0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623    -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920   -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108   -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329   -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[492 rows x 31 columns]

**CONVERTING TO BALANCED DATA USING SAMPLING**

In [ ]:
valid_sample=valid.sample(n=len(fraud))
valid_sample

Time        V1        V2        V3        V4        V5        V6  \
155615  105904.0  1.972265 -0.464781 -1.626202 -0.593907  1.901315  3.742424   
58327    48332.0  1.159757  0.222481  0.237549  1.089785 -0.138051 -0.498851   
268331  163157.0  1.932350 -0.466472 -0.453639  0.166694 -0.540701 -0.302526   
118896   75249.0 -1.142650  0.418000  1.472538 -1.084475 -0.260661 -0.992462   
184416  126233.0 -1.329959 -0.123651  0.933031 -3.086828 -0.494215 -0.546785   
...          ...       ...       ...       ...       ...       ...       ...   
106073   69810.0 -1.375364  1.228114  1.136320 -0.005838  0.048109 -0.472369   
71928    54499.0  1.106163  0.251623  0.509677  1.354759 -0.465662 -1.021819   
158492  111338.0 -1.787262 -1.272808  1.992868 -1.918383  0.814814 -0.931105   
280325  169456.0 -1.806268  2.117882 -0.549426  0.806450  0.230671 -0.236108   
274051  165830.0  2.249375 -1.506309 -0.753225 -1.659968 -1.345344 -0.325434   

              V7        V8        V9  ...       V21       V22       V23  \
155615 -1.173426  0.927211  2.441938  ... -0.265432 -0.511603  0.396600   
58327   0.143770 -0.079237  0.032431  ...  0.063782  0.199615 -0.083720   
268331 -0.510143  0.044381  0.782174  ... -0.215570 -0.610643  0.366772   
118896  0.228642  0.204537 -0.305246  ...  0.061036  0.011380 -0.055293   
184416 -0.311817  0.351724 -2.682353  ... -0.142919  0.014963 -0.164697   
...          ...       ...       ...  ...       ...       ...       ...   
106073  0.498308  0.118234  0.050762  ... -0.084563 -0.521705 -0.011793   
71928   0.269689 -0.284017 -0.175024  ...  0.134855  0.355890 -0.127290   
158492 -0.789848  0.246974  0.338113  ... -0.075656 -0.559358  0.112481   
280325  0.382590  1.049429 -1.733320  ...  0.242636  0.352500 -0.560554   
274051 -1.351221 -0.038003 -1.235448  ... -0.119383  0.129000  0.182379   

             V24       V25       V26       V27       V28  Amount  Class  
155615  0.583540 -0.353651 -0.943001  0.050793 -0.038305    9.39      0  
58327   0.077463  0.594408 -0.299754  0.023782  0.019868   25.65      0  
268331 -0.345914 -0.644328  0.229921 -0.048490 -0.052194   34.99      0  
118896  0.580368 -0.664077  0.627306 -0.359990 -0.212035   23.71      0  
184416 -0.333863  0.897662 -0.106190  0.047456 -0.004645   15.00      0  
...          ...       ...       ...       ...       ...     ...    ...  
106073 -0.011151 -0.148345 -0.840765 -0.926287 -0.239212    9.99      0  
71928   0.754913  0.639954 -0.310024  0.014061  0.036820   58.90      0  
158492 -0.458250  0.493589 -0.465938  0.160906  0.126860   68.00      0  
280325 -0.428000  1.182923 -0.199254 -0.581635 -0.183680   22.88      0  
274051 -0.516507 -0.259402 -0.166645  0.019391 -0.058052   25.20      0  

[492 rows x 31 columns]

**Rechecking for Balanced Data**

In [ ]:
print(valid_sample.shape)
print(fraud.shape)

(492, 31)
(492, 31)


hence the data is balanced.Hence combining both Valid & Fraud Users to form a new dataset

In [ ]:
new_Dataset=pd.concat([valid_sample,fraud],axis=0)
new_Dataset

Time        V1        V2        V3        V4        V5        V6  \
155615  105904.0  1.972265 -0.464781 -1.626202 -0.593907  1.901315  3.742424   
58327    48332.0  1.159757  0.222481  0.237549  1.089785 -0.138051 -0.498851   
268331  163157.0  1.932350 -0.466472 -0.453639  0.166694 -0.540701 -0.302526   
118896   75249.0 -1.142650  0.418000  1.472538 -1.084475 -0.260661 -0.992462   
184416  126233.0 -1.329959 -0.123651  0.933031 -3.086828 -0.494215 -0.546785   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
155615 -1.173426  0.927211  2.441938  ... -0.265432 -0.511603  0.396600   
58327   0.143770 -0.079237  0.032431  ...  0.063782  0.199615 -0.083720   
268331 -0.510143  0.044381  0.782174  ... -0.215570 -0.610643  0.366772   
118896  0.228642  0.204537 -0.305246  ...  0.061036  0.011380 -0.055293   
184416 -0.311817  0.351724 -2.682353  ... -0.142919  0.014963 -0.164697   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
155615  0.583540 -0.353651 -0.943001  0.050793 -0.038305    9.39      0  
58327   0.077463  0.594408 -0.299754  0.023782  0.019868   25.65      0  
268331 -0.345914 -0.644328  0.229921 -0.048490 -0.052194   34.99      0  
118896  0.580368 -0.664077  0.627306 -0.359990 -0.212035   23.71      0  
184416 -0.333863  0.897662 -0.106190  0.047456 -0.004645   15.00      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [ ]:
new_Dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
155615  105904.0  1.972265 -0.464781 -1.626202 -0.593907  1.901315  3.742424   
58327    48332.0  1.159757  0.222481  0.237549  1.089785 -0.138051 -0.498851   
268331  163157.0  1.932350 -0.466472 -0.453639  0.166694 -0.540701 -0.302526   
118896   75249.0 -1.142650  0.418000  1.472538 -1.084475 -0.260661 -0.992462   
184416  126233.0 -1.329959 -0.123651  0.933031 -3.086828 -0.494215 -0.546785   

              V7        V8        V9  ...       V21       V22       V23  \
155615 -1.173426  0.927211  2.441938  ... -0.265432 -0.511603  0.396600   
58327   0.143770 -0.079237  0.032431  ...  0.063782  0.199615 -0.083720   
268331 -0.510143  0.044381  0.782174  ... -0.215570 -0.610643  0.366772   
118896  0.228642  0.204537 -0.305246  ...  0.061036  0.011380 -0.055293   
184416 -0.311817  0.351724 -2.682353  ... -0.142919  0.014963 -0.164697   

             V24       V25       V26       V27       V28  Amount  Class  
155615  0.583540 -0.353651 -0.943001  0.050793 -0.038305    9.39      0  
58327   0.077463  0.594408 -0.299754  0.023782  0.019868   25.65      0  
268331 -0.345914 -0.644328  0.229921 -0.048490 -0.052194   34.99      0  
118896  0.580368 -0.664077  0.627306 -0.359990 -0.212035   23.71      0  
184416 -0.333863  0.897662 -0.106190  0.047456 -0.004645   15.00      0  

[5 rows x 31 columns]

In [ ]:
new_Dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [ ]:
new_Dataset.shape

(984, 31)

**Rechecking for Null Values**

In [ ]:
new_Dataset.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
new_Dataset["Class"].value_counts()

0    492
1    492
Name: Class, dtype: int64

Hence the data is balanced.Now,
* Splitting the data into 2 variables ---x & y taking target function in y--variable and data in x--variable

Using drop method

In [ ]:
x=new_Dataset.drop("Class",axis=1)
x

Time        V1        V2        V3        V4        V5        V6  \
155615  105904.0  1.972265 -0.464781 -1.626202 -0.593907  1.901315  3.742424   
58327    48332.0  1.159757  0.222481  0.237549  1.089785 -0.138051 -0.498851   
268331  163157.0  1.932350 -0.466472 -0.453639  0.166694 -0.540701 -0.302526   
118896   75249.0 -1.142650  0.418000  1.472538 -1.084475 -0.260661 -0.992462   
184416  126233.0 -1.329959 -0.123651  0.933031 -3.086828 -0.494215 -0.546785   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
155615 -1.173426  0.927211  2.441938  ... -0.189518 -0.265432 -0.511603   
58327   0.143770 -0.079237  0.032431  ... -0.156109  0.063782  0.199615   
268331 -0.510143  0.044381  0.782174  ... -0.114649 -0.215570 -0.610643   
118896  0.228642  0.204537 -0.305246  ... -0.112473  0.061036  0.011380   
184416 -0.311817  0.351724 -2.682353  ... -0.316675 -0.142919  0.014963   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
155615  0.396600  0.583540 -0.353651 -0.943001  0.050793 -0.038305    9.39  
58327  -0.083720  0.077463  0.594408 -0.299754  0.023782  0.019868   25.65  
268331  0.366772 -0.345914 -0.644328  0.229921 -0.048490 -0.052194   34.99  
118896 -0.055293  0.580368 -0.664077  0.627306 -0.359990 -0.212035   23.71  
184416 -0.164697 -0.333863  0.897662 -0.106190  0.047456 -0.004645   15.00  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [ ]:
y=new_Dataset.Class
y

155615    0
58327     0
268331    0
118896    0
184416    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

Data Standardization

In [ ]:
scaler=StandardScaler()
scaler.fit(x)

StandardScaler()

In [ ]:
Standardised_data=scaler.transform(x)
Standardised_data

array([[ 0.36963619,  0.78078391, -0.5742262 , ..., -0.01730226,
        -0.18342115, -0.24808736],
       [-0.83314394,  0.63435341, -0.39766899, ..., -0.04379108,
        -0.04330238, -0.21199302],
       [ 1.56575185,  0.77359053, -0.5746607 , ..., -0.1146682 ,
        -0.21687591, -0.19125986],
       ...,
       [ 1.69515538,  0.30348695, -0.16546171, ...,  0.31055595,
         0.37699723, -0.09602942],
       [ 1.70800381, -0.13583429, -0.30431624, ...,  0.80067245,
        -0.70224105,  0.27492534],
       [ 1.71598446,  0.78433633, -0.41411197, ..., -0.06418437,
        -0.12803178, -0.17452239]])

Now taking this standardised data as x value because it has to be suitable for all the data

In [ ]:
x=Standardised_data

In [ ]:
print(x)

[[ 0.36963619  0.78078391 -0.5742262  ... -0.01730226 -0.18342115
  -0.24808736]
 [-0.83314394  0.63435341 -0.39766899 ... -0.04379108 -0.04330238
  -0.21199302]
 [ 1.56575185  0.77359053 -0.5746607  ... -0.1146682  -0.21687591
  -0.19125986]
 ...
 [ 1.69515538  0.30348695 -0.16546171 ...  0.31055595  0.37699723
  -0.09602942]
 [ 1.70800381 -0.13583429 -0.30431624 ...  0.80067245 -0.70224105
   0.27492534]
 [ 1.71598446  0.78433633 -0.41411197 ... -0.06418437 -0.12803178
  -0.17452239]]


In [ ]:
print(y)

155615    0
58327     0
268331    0
118896    0
184416    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Splitting the data into train_data (80%) and test_data(20%) and taking shuffle_state nothing but random_state as 3

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=3)

In above test_size means 20% of test_data

Printing the shapes of train_data and test_data

In [ ]:
print(x.shape,x_train.shape,x_test.shape)

(984, 30) (787, 30) (197, 30)


In [ ]:
print(y.shape,y_train.shape,y_test.shape)

(984,) (787,) (197,)


**Using SVM Model**

Training the model

In [ ]:
classifier=svm.SVC(kernel="linear")

SVC--Support Vector Classifier

In [ ]:
#train the support vector machine classifier
classifier.fit(x_train,y_train)

SVC(kernel='linear')

In [ ]:
print(classifier)

SVC(kernel='linear')


**Model Evaluation**

* based on accuracy score
* 2 ways

 * making prediction for the same data i.e test-data
 * making prediction for new data

In [ ]:
#finding accuracy_score for the train_data
x_train_prediction= classifier.predict(x_train)
training_data_accuracy= accuracy_score(x_train_prediction, y_train)
print("Accuracy score of Train Data: ", training_data_accuracy)

Accuracy score of Train Data:  0.9479034307496823


In [ ]:
#finding accuracy_score for the test_data
x_test_prediction=classifier.predict(x_test)
test_data_accuracy= accuracy_score(x_test_prediction, y_test)
print("Accuracy score of Test Data: ", test_data_accuracy)

Accuracy score of Test Data:  0.949238578680203


Making Prediction

In [ ]:
input_data=[172792,-0.533412522,-0.189733337,0.703337367,-0.50627124,-0.012545679,-0.649616686,1.577006254,-0.414650408,0.486179505,-0.915426649,-1.040458335,-0.031513054,-0.188092901,-0.08431647,0.041333455,-0.302620086,-0.660376645,0.167429934,-0.256116871,0.382948105,0.261057331,0.643078438,0.376777014,0.008797379,0.473648704,-0.818267121,-0.002415309,0.013648914,217]

#converting it into numpy array
input_data_as_numpy_array=np.array(input_data)
#reshaping the array since we are predicting
input_data_reshaped= input_data_as_numpy_array.reshape(1,-1)
#Standardise  the input data
std_data=scaler.transform(input_data_reshaped)
print(std_data)
prediction=classifier.predict(std_data)
print(prediction)
if prediction[0]==0:
  print("The Credit Card is valid")
elif prediction[0]==1:
  print("The Credit Card is fraud")

[[ 1.76704391  0.32920983 -0.50356666  0.67766529 -0.87957947  0.37525349
   0.04467089  0.73683203 -0.13873771  0.77293401  0.42828061 -1.06728562
   0.67393947 -0.10865253  0.72900167  0.11026122  0.50510299  0.44511687
   0.52542181 -0.4249678   0.12484866 -0.03808747  0.52154529  0.34488034
   0.09121145  0.65338384 -1.7849437  -0.06948292 -0.05828183  0.21277027]]
[0]
The Credit Card is valid


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


Making Prediction

**CONCLUSION**
* The accuracy of Test Data using SVM is 94.9%
* Hence the test result and prediction result are  more accurate hence we can say that the SVM model is more optimal.